### PR TITLE
chore!: Rewrite Release Workflow to Github Workflows entirely

### PR DESCRIPTION
**Pull request by oliverbaehler**
With this Pull Request the entire Workflow in this repository is changed. Let me take you through the changes, what they are causing etc.

But first of all, why not considering #565? Because there's already too many unrelated changes in there. But the bigger problem, it does not resolve some of the key problems with the current release workflow. It's really ugly that we push chart artifacts to the gh-pages branch. In my opinion they don't belong there but rather each new helm release gets it's own Github release and the archive appended to that github release. That's how almost all of the people are doing it, that are hosting charts on Github. And also these scripts are not really maintainable for the future. Nowadays we have solutions that already provide the functionality we need to release charts. With this I am poiting towards [chart-releaser](https://github.com/helm/chart-releaser) and [chart-testing](https://github.com/helm/chart-testing) which are both projects which were exactly build for this purpose. So that's why i propose moving to those technologies and abandon these old scripts.

The future Workflow will need look much different than it's now. After a User from a forked repository creates a Pull Request the following two actions are triggered:

  1. Helm Chart Linting & Testing Action.
     a. The linting checks for all different kind of semantic errors as well as the old lintconf is enforced for yaml styles. The linting is executed on all charts, which have any changes in their directory. The lint check also validates if a version bump was done.
     b. If the lint was successful, the changed charts will then be install into a KinD Cluster to see if it's possible to install them with default values or if they cause any errors.

   2. Pull Request Size labeling
     a. This is something known from the helm projects. The Pull Request will be labeled with the Size depending of how many lines of code were changed. This makes it already clear by the label how complex a pr could be and might encourage more approvals from users :+1:  

And ofc other actions which are not managed by github workflows like DCO etc.
These workflows have to succeed to be able to merge the pr. Once the changes are pushed the master, the following happens:

   1. Chart Publish Action
      a. All charts which contain changes against the latest github tag are release. Here it's important that the current index is previously downloaded and then used to be merged (see ct.yaml). Otherwise the index would be generated based on new changes and existing tags. For each chart there's a new github release made and the chart binary appended. Finally the index is uploaded to the gh-pages branch.

And that concludes the new workflow.  To not have any problems in the future we will create for each chart a new bugfix release instantly, so that all charts have a new release as github release. I have tested the workflow on my fork and it worked for me. The most important thing for me, was that the old charts still are available and that's also the case. So we don't have to do any manual transition, everything should just work :).

I have removed the old scripts and changed the lint script that it also works locally. The reviewer should probably test it as well, to see how it feels like. But I think it's already a good improvements. We could also talk about signing our releases in the future, but one step after another.

Checklist:

* [x] I have updated the chart version in `Chart.yaml` following Semantic Versioning.
* [x] Any new values are backwards compatible and/or have sensible default.
* [x] I have followed the testing instructions in the [contributing guide](https://github.com/argoproj/argo-helm/blob/master/CONTRIBUTING.md).
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo).
* [ ] My build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

Changes are automatically published when merged to `master`. They are not published on branches.

### CODE REVIEW

**Comment by oliverbaehler** [comment]
As you can see, the Checks are running through already: https://github.com/argoproj/argo-helm/pull/652/checks. So you can kinda guess what it would look like in the future. After this is merged we could remove Circle Ci completely, Right now it's failing bcs there is no more configuration (https://app.circleci.com/pipelines/github/argoproj/argo-helm/2073/workflows/f4dc64b0-713c-4a32-9525-b0c28e09a2ea/jobs/3433).

I would be happy if we could merge this asap, otherwise I will always have to rebase changes etc. 

I guess the Repository Lords should also take a look at this. @jessesuen @alexmt @alexec 

PS: If you are looking for additional repository owners I would love to help contribute. Currently I can not really speed up things since I am not able to merge anything or so, bcs i don't have write access to the repository. If not that's okay as well.

**Comment by alexec** [comment]
I don’t feel strongly about this. 

I would propose to break up the PR into non-controversial lint and migrating to GIthub Actions in one PR and publishing into another.

Could we have boron GitHub releases AND gh-pages?

**Review by mkilchhofer — COMMENTED** [review]
For me this looks okay and I like GH actions for chart publishing.
I'd proceed if our goal is to get rid of the CircleCI workflow in the future.

Added just a few nits.

File: .github/workflows/lint-and-test.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
+      - name: Set up Helm
+        uses: azure/setup-helm@v1
+
+      - uses: actions/setup-python@v2
```
Comment (mkilchhofer) [inline .github/workflows/lint-and-test.yml:17]: Maybe add a friendly name inside the pipeline for this?
```suggestion
      - name: Setup python
        uses: actions/setup-python@v2
```

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+      - name: Configure Git
+        run: |
+          git config user.name "$GITHUB_ACTOR"
+          git config user.email "$GITHUB_ACTOR@users.noreply.github.com"  
+
+      ## This is required to consider the old Circle-CI Index and to stay compatible with all the old releases.
+      - name: Fetch current Chart Index 
+        run: |
+          git checkout origin/gh-pages index.yaml
```
[the remark this answers is not in the document]
Comment (oliverbaehler) [reply .github/workflows/publish.yml:38]: Yep, Otherwise it will just throw away the old index and I guess you should still be able to pull old charts?

**Comment by oliverbaehler** [comment]
@alexec Sure I can break it up into two prs, but it doesn't mkae it simplier.

> Could we have boron GitHub releases AND gh-pages?

Why would you want that?

File: .github/workflows/publish.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mkilchhofer) [reply .github/workflows/publish.yml:38]: Are u sure? Why would the new action throw away this file? Also this action uses this file in exactly this branch. Do I miss something?

**Comment by mkilchhofer** [comment]
Closed as @oliverbaehler filed a new PR with reduced scope: chore!: #741

**Pull request closed without merging**